### ISSUE TITLE
Support for Wemo Outdoor Plug

### ISSUE THREAD

**Issue by carterthayer** [pywemo/pywemo#177]
Discovery and manual input for this device does not work with this device

**Comment by pavoni** [comment]
Do you have any more info? Do you get any error traces?

It's quite possible that Belkin are moving away from their local control model - as suggested here:-

https://github.com/pavoni/pywemo/issues/171#issuecomment-721463048

If that's the case then this pywemo may be unable to control new devices.

**Comment by carterthayer** [comment]
No errors. It was not found when doing discovery and the device object is None when using the IP method from the README.

I will do some more digging to figure out why. Do you have any guidance/recommendations for looking at what is happening with new devices?

**Comment by pavoni** [comment]
I've seen some discussions about Belkin now requiring a cloud account registration - and moving from a local network model of control to a cloud one. So it's possible that the new device doesn't support the old style api.

If the device will talk to you there are parts of the code here can help in getting service names etc. but if you can't even get the device to talk to you - then the they only way is to use something like wireshark to see what is happening on the network between the Wemo app and the device as @spectre5  is doing here https://github.com/pavoni/pywemo/issues/171 but that is not an easy job!

**Comment by Spectre5** [comment]
When the commands go to/from the cloud, they are encrypted so figuring out what the commands are from the app will be quite difficult, if not impossible.  With wireshark you'll be able to see the traffic, but to decrypt that message you'll need the decryption key.  This is possible in some cases where Firefox can save the key to a file for you, as described
[here](https://unit42.paloaltonetworks.com/wireshark-tutorial-decrypting-https-traffic/) or [here](https://www.comparitech.com/net-admin/decrypt-ssl-with-wireshark/).  However, I don't know how you would get the decryption key from the wemo app.  Maybe there is some android app that could intercept and output it for you, but I'm not aware of any.

Basically, I wouldn't be surprised if new and future Wemo devices don't work with pywemo.  With that said, I'd be a little bit surprised if the entire protocol is different for the outdoor plug, since I have a few switches on the latest firmware and pywemo still works fine.  So before you give up on this outdoor plug, you should debug it some more.

- Are you sure you're connect to the same network?
- Can you ping in at the IP?
- Try nmap to see what ports are open on it.
- Fire up wireshark and see what you can see, I imagine that some of the communication between the app and device is still local.

**Comment by mattbilodeau** [comment]
I was able to get details from my outdoor plug, via netdisco:

{'host': '192.168.1.7',
  'mac_address': '{mac}',
  'manufacturer': 'Belkin International Inc.',
  'model_name': 'OutdoorPlug',
  'model_number': '1.0',
  'name': 'Ice Melter',
  'port': 49152,
  'serial': '{Serial}',
  'ssdp_description': 'http://192.168.1.7:49152/setup.xml',
  'udn': 'uuid:OutdoorPlug-1_0-{Serial}',
  'upnp_device_type': 'urn:Belkin:device:outdoor:1'},

then browsing to the ssdp_description, i think the state/control area is:
            <service>
                <serviceType>urn:Belkin:service:basicevent:1</serviceType>
                <serviceId>urn:Belkin:serviceId:basicevent1</serviceId>
                <controlURL>/upnp/control/basicevent1</controlURL>
                <eventSubURL>/upnp/event/basicevent1</eventSubURL>
                <SCPDURL>/eventservice.xml</SCPDURL>
            </service>

Then using details from https://alexmeub.com/controlling-wemo-with-esp8266/

gets the state of my switch
POST: http://<WEMO-IP>:<port>/upnp/control/basicevent1
Content-Type: text/xml; charset="utf-8"
SOAPACTION: "urn:Belkin:service:basicevent:1#GetBinaryState"

<?xml version="1.0" encoding="utf-8"?>
<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/">
    <s:Body>
        <u:GetBinaryState xmlns:u="urn:Belkin:service:basicevent:1">
        </u:GetBinaryState>
    </s:Body>
</s:Envelope>

and this properly sets the device state.

POST: http://<WEMO-IP>:<port>/upnp/control/basicevent1
Content-Type: text/xml; charset="utf-8"
SOAPACTION: "urn:Belkin:service:basicevent:1#SetBinaryState"

<?xml version="1.0" encoding="utf-8"?>
<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/">
    <s:Body>
        <u:SetBinaryState xmlns:u="urn:Belkin:service:basicevent:1">
        	<BinaryState>1</BinaryState>
        </u:SetBinaryState>
    </s:Body>
</s:Envelope>

It appears that the POST requests are a slightly different configuration than the Switch Post/Get Requests.  

Since the uuid:OutdoorPlug doesn't exist in the device_from_uuid_and_location, it doesn't return from a scan.

------------------------
modifying discovery.py on line 13:
from .ouimeaux_device.outdoorplug import OutdoorPlug

and adding on line 73:
    if uuid.startswith('uuid:OutdoorPlug'):
        return OutdoorPlug(url=location, mac=mac, rediscovery_enabled=rediscovery_enabled)

creating a copy of ouimeaux_device/switch.py  and saving it to ouimeaux_device/outdoorplug.py
Modifying the class name from Switch to OutdoorPlug
changing switch to OutdoorPlug on 29 and 34

that will get us close, I'm out of time for tonight to compare the Switch POSTs and the OutdoorPlug posts, but this is close.

**Comment by Spectre5** [comment]
This is great detective work @mattbilodeau.  @pavoni, this is something we probably both overlooked with our first responses --- since the outdoor plug doesn't exist in library, `device_from_uuid_and_location` returns `None`, so the device is never "discovered".  I wonder if it makes more sense to `return Device(url=location, mac=mac, rediscovery_enabled=rediscovery_enabled)` in the case where the uuis is not recognized?  This would allow some future Wemo device to still be discovered, just not directly usable since it wouldn't have the helper methods like `on`, `off`, etc defined for it.  But at least it would let a user know that something was discovered on the network.  That said, I'm not sure if there would be downstream effects to this (i.e. in Home Assistant).  We could make a new class named `UnknownDevice` that is just identical to `Device` and which would then allow a downstream library to check against it.  So, for example, perhaps Home Assistant would want to ignore devices that return `True` for `isinstance(object, UnknownDevice)` since it likely cannot be controlled, but it would still allow someone using `pywemo` to find the device and interact with it to work towards creating an official device class for it.

**Comment by Spectre5** [comment]
@mattbilodeau, can you please check out and test the code in #184.  I've hopefully added basic support for it there based on your findings, but I can't test it myself.

**Comment by jeffehobbs** [comment]
Hey, I can return early success, with the discovery portion at least. Running `pywemo.discover_devices()` on the current, production version of pywemo did not return the outdoor plug for me. After updating to the code in #184, I now get `<WeMo OutdoorPlug "Outdoor Christmas Lights">` in the array of devices returned. Really hopeful! 🎅

**Comment by mattbilodeau** [comment]
I agree with @jeffehobbs, and i was able to toggle the outlet on and off with pywemo!  thanks for the quick response.  


```
import pywemo
devices = pywemo.discover_devices()
print(devices)

for device in devices:
    if device.host == "192.168.1.7":
        print(device.host)
        print("current state:")
        print(device.get_state())
        device.on()
        print("state after on()")
        print(device.get_state())
        device.off()
        print("state after off()")
        print(device.get_state())
        device.toggle()
        print("state after toggle()")
        print(device.get_state())](url)
```


generated:

```
matt@RedDelicious pywemo % python3 wemo.py
[<WeMo OutdoorPlug "Ice Melter">, <WeMo Insight "Dehumid">, <WeMo Switch "Light">, <WeMo LightSwitch "BD">, <WeMo Switch "Lamp">, <WeMo Switch "Tree">]
192.168.1.7
current state:
1
state after on()
1
state after off()
0
state after toggle()
1
```

**Comment by pavoni** [comment]
Sounds like it's time to merge!
